src/xref2/tools.ml

```diff
@@ -6,11 +6,87 @@ open ResultMonad
 
 type ('a, 'b) either = Left of 'a | Right of 'b
 
+let filter_map f x =
+  List.rev
+  @@ List.fold_left
+       (fun acc x -> match f x with Some x -> x :: acc | None -> acc)
+       [] x
+
 type module_modifiers =
   [ `Aliased of Cpath.Resolved.module_ | `SubstMT of Cpath.Resolved.module_type ]
 
 type module_type_modifiers = [ `AliasModuleType of Cpath.Resolved.module_type ]
 
+(* These three functions take a fully-qualified canonical path and return
+   a list of shorter possibilities to test *)
+let c_mod_poss env p =
+  (* canonical module paths *)
+  let rec inner = function
+    | `Dot (p, n) -> (
+        let rest = List.map (fun p -> `Dot (p, n)) (inner p) in
+        match Env.lookup_by_name Env.s_module n env with
+        | Ok (`Module (id, m)) ->
+            let m = Component.Delayed.get m in
+            `Identifier (id, m.hidden) :: rest
+        | Error _ -> rest)
+    | p -> [ p ]
+  in
+  inner p
+
+let c_modty_poss env p =
+  (* canonical module type paths *)
+  match p with
+  | `Dot (p, n) -> (
+      let rest = List.map (fun p -> `Dot (p, n)) (c_mod_poss env p) in
+      match Env.lookup_by_name Env.s_module_type n env with
+      | Ok (`ModuleType (id, _)) -> `Identifier (id, false) :: rest
+      | Error _ -> rest)
+  | p -> [ p ]
+
+let c_ty_poss env p =
+  (* canonical type paths *)
+  match p with
+  | `Dot (p, n) -> (
+      let rest = List.map (fun p -> `Dot (p, n)) (c_mod_poss env p) in
+      match Env.lookup_by_name Env.s_type n env with
+      | Ok (`Type (id, _)) ->
+          `Identifier ((id :> Odoc_model.Paths.Identifier.Path.Type.t), false)
+          :: rest
+      | Error _ -> rest)
+  | p -> [ p ]
+
+(* Small helper function for resolving canonical paths.
+   [canonical_helper env resolve lang_of possibilities p2] takes the
+   fully-qualified path [p2] and returns the shortest resolved path
+   whose identifier is the same as the resolved fully qualified path.
+   [resolve] is a function that resolves an arbitrary unresolved path,
+   [lang_of] turns a resolved path into a generic resolved Lang path
+   and [possibilities] is a function that, given the fully qualified
+   unresolved path, returns a list of all possible unresolved paths
+   (including the longest one) *)
```

Comment:
`possibilities` should be sorted, shortest first ?

Reply:
Yes, I'll mention that.